src/limitations/prerelease_versions.md

```diff
@@ -0,0 +1,64 @@
+# Pre-release versions
+
+Pre-releasing is a very common pattern in the world of versioning.
+It is however one of the worst to take into account in a dependency system, and I highly recommend that if you can avoid introducing pre-releases in your package manager, you should.
+In the context of pubgrub, pre-releases break two fondamental properties of the solver.
+
+1. Pre-releases act similar to continuous spaces.
+2. Pre-releases break the mathematical properties of subsets in a space with total order.
+
+(1) Indeed, it is hard to answer what version comes after "1-alpha0".
+Is it "1-alpha1", "1-beta0", "2"?
+In practice, we could say that the version that comes after "1-alpha0" is "1-alpha0?" where the "?" character is chosen to be the lowest character in the lexicographic order, but we clearly are on a stretch here and it certainly isn't natural.
+
+(2) Pre-releases are often semantically linked to version constraints written by humans, interpreted differently depending on context.
+For example, "2.0.0-beta" is meant to exist previous to version "2.0.0".
+Yet, it is not supposed to be contained in the set described by `1.0.0 <= v < 2.0.0`, and only within sets where one of the bounds contains a pre-release marker such as `2.0.0-alpha <= v < 2.0.0`.
+This poses a problem to the dependency solver because of backtracking.
+Indeed, the PubGrub algorithm relies on knowledge accumulated all allong the propagation of the solver front.
+And this knowledge is composed of facts, that are thus never backtracked even when backtracking happens.
+Those facts are called incompatibilities and more info about those is available in the "Internals" section of the guide.
+The problem is that if a fact recalls that there is no version within the `1.0.0 <= v < 2.0.0` range, backtracking to a situation where we ask for a version within `2.0.0-alpha <= v < 2.0.0` will return nothing even without checking if a pre-release exists in that range.
+And this is one of the fundamental mechanisms of the algorithm, so we should not try to alter it.
+
+## Playing again with packages?
+
+In the light of the "bucket" and "proxis" scheme we introduced in the section about allowing multiple versions per package, I'm wondering if we could do something similar for pre-releases.
+Normal versions and pre-release versions would be split into two subsets, each attached to a different bucket.
+This is probably not going to work because of places in the algorithm where intersections of ranges are computed.
+In those places, only one of the two buckets will have the intersection applied.
+So if we intersect `2.0.0-alpha <= v < 2.0.0` with `2.0.0 <= v < 3.0.0`, we want the result to be empty, but with a bucket scheme, I fear that the pre-release bucket will not be concerned by the intersection and we will still have pre-releases within `2.0.0-alpha <= v < 2.0.0` as still valid.
```

Comment:
For Cargo, we can not put `2.0.0-alpha` and `2.0.0` in different buckets, because they can not be activated at the same time. Unless we add some way for a DependacyProveder to add incompatibility directly or some way to have a negative dependency (I depend on that not being selected), but that is also an API braking change.

Reply:
Would it break pubgrub if we say: "a" depends on the prerelease bucket of "b" in range Range::none()? Or would it just make the pre-release bucket not pickable? I don't remember the exact semantics

Reply:
I may have a shot at it just to convince myself

Reply:
> some way to have a negative dependency

`Term` is already in the public API since it's needed for error reporting if I remember well. So I think all it would take would be to have `get_dependencies` return terms instead of ranges. Most of the time you'd use `Positive(range)` but a scenario like this could take advantage of `Negative(range)`. Maybe that's the smallest change we need to enable pre-releases? Of course we'd have to check how that impacts all the rest of the implementations, like how it impacts buckets/proxies for multiple versions etc.